### PR TITLE
Update source link page with deterministic builds

### LINKED ISSUE #17710: Recommend deterministic builds (Open-source library guidance)
Deterministic builds are now available, either using the preview SDK, or [with a minor hack via the existing SDK](https://github.com/dotnet/sourcelink/issues/572).

Should the recommendation be updated to recommend deterministic builds? or should this be deferred until SDK 3.1.300 ships? Note that guidance on how [exists here](https://github.com/clairernovotny/DeterministicBuilds)

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: e80fe2fd-cbdb-699e-0dbb-a0a310bd372c
* Version Independent ID: 266d08ae-344a-683b-11e2-143b704e21a8
* Content: [Open-source .NET library guidance](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/)
* Content Source: [docs/standard/library-guidance/index.md](https://github.com/dotnet/docs/blob/master/docs/standard/library-guidance/index.md)
* Product: **dotnet**
* GitHub Login: @JamesNK
* Microsoft Alias: **jamesnk**

### LINKED ISSUE #17735: Consider adding guidance regarding deterministic builds
As demonstrated by @clairernovotny [here](https://github.com/clairernovotny/DeterministicBuilds), it is likely recommended to "CONSIDER" deterministic builds if one considers using Source Link.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 5db5ae44-5ee9-ed62-7355-7c0cbc1bfc07
* Version Independent ID: f54c98e2-bbae-42c6-2103-e94d6ec0e931
* Content: [Source Link and .NET libraries](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/sourcelink)
* Content Source: [docs/standard/library-guidance/sourcelink.md](https://github.com/dotnet/docs/blob/master/docs/standard/library-guidance/sourcelink.md)
* Product: **dotnet**
* GitHub Login: @JamesNK
* Microsoft Alias: **jamesnk**

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/dotnet/docs/issues/17735
Fixes https://github.com/dotnet/docs/issues/17710

I think longer term we would want to link to instructions provided by the Source Link or NuGet repo, but https://github.com/clairernovotny/DeterministicBuilds seems fine to me for now.

@clairernovotny @tkp1n @mgravell

### CODE REVIEW

**Comment by clairernovotny** [comment]
@JamesNK do you want the repo to be in my user account or would it be better to put it somewhere else (not sure where)?

**Comment by mgravell** [comment]
The logical "somewhere else" would be "here in detail". I don't have a dog in the race of "here vs there".

**Comment by JamesNK** [comment]
> @JamesNK do you want the repo to be in my user account or would it be better to put it somewhere else (not sure where)?

Should the discussion and instructions around deterministic builds be added to the Source Link repo? Your repo is fine. I don't feel strongly about it.

**Comment by clairernovotny** [comment]
We should probably have more instructions there...and probably better verbiage around reasons for deterministic builds. What I had was done in 30 seconds trying to whip up a sample.

/cc @jaredpar @tmat

File: docs/standard/library-guidance/sourcelink.md [hunk 1]
```diff
@@ -32,6 +32,10 @@ You can use [NuGet Package Explorer](https://github.com/NuGetPackageExplorer/NuG
 
 > For the best debugging experience your library should publish symbol files as well as use Source Link. For more information about symbol files and symbol packages, see [Symbol packages](./nuget.md#symbol-packages).
 
+✔️ CONSIDER enabling deterministic builds.
+
+> Deterministic builds enable verification that the resulting binary was built from the specified source and provides traceability. For more information about deterministic builds, and instructions for enabling them, see [Deterministic Builds](https://github.com/clairernovotny/DeterministicBuilds).
```
Comment (pkulikov) [inline docs/standard/library-guidance/sourcelink.md:37]:
```suggestion
> Deterministic builds enable verification that the resulting binary was built from the specified source and provide traceability. For more information about deterministic builds and instructions for enabling them, see [Deterministic Builds](https://github.com/clairernovotny/DeterministicBuilds).
```

Comment (jaredpar) [reply docs/standard/library-guidance/sourcelink.md:37]: > Deterministic builds enable verification that the resulting binary was built from the specified source and provides traceability.

What tool is providing this verification?

Comment (clairernovotny) [reply docs/standard/library-guidance/sourcelink.md:37]: We don't have a tool yet to do the verification. Getting the Source Link and deterministic settings are a prerequisite for any tool to work.

**Comment by tmat** [comment]
@clairernovotny From the sample repo readme:

> Deterministic builds require two properties set to true during CI: ContinuousIntegrationBuild and Deterministic. These should not be enabled during local dev or the debugger won't be able to find the local source files.

`Deterministic` should be enabled always, and it's in fact enabled by default in .NET SDK projects. `Deterministic` itself won't trigger path mapping. `DeterministicSourcePaths` is a property that controls the mapping. Setting `ContinuousIntegrationBuild` to true will result in `DeterministicSourcePaths` enabled (if `Deterministic` is also enabled).

tldr: The only property that should be set in CI and not locally is `ContinuousIntegrationBuild`.

**Comment by tmat** [comment]
It'd also be useful to explain why `GetPathMap` target is needed for now:
https://github.com/clairernovotny/DeterministicBuilds/blob/master/Directory.Build.targets#L15

**Comment by tmat** [comment]
> Should the discussion and instructions around deterministic builds be added to the Source Link repo?

I'd prefer it to be in dotnet/docs. Source Link is just one component that participates. Source Link would then link to dotnet/docs.

**Comment by clairernovotny** [comment]
@tmat I just updated the sample. removed the `Deterministic` property and the extra Target. Coverlet's instructions will include the proper target to add for their workaround.

**Comment by gewarren** [comment]
@JamesNK It looks like there is some feedback to incorporate. Are you still working on this PR?

**Comment by JamesNK** [comment]
I'm happy with the PR as it currently is. Feel free to approve and merge.

**Review by gewarren — APPROVED** [review]
